cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -0,0 +1,868 @@
+  std::vector<std::future<thrust::host_vector<bool>>> data_page_mask_tasks;
+  data_page_mask_tasks.reserve(num_columns);
+  auto streams = cudf::detail::fork_streams(stream, num_columns);
+
+  // For all columns, look up which pages contain at least one required row. i.e.
+  // !validity_it[row_idx] or is_row_required[row_idx] satisfies, and add its byte range to the
+  // output list of byte ranges for the column.
+  std::for_each(
+    thrust::counting_iterator<size_t>(0),
+    thrust::counting_iterator(num_columns),
+    [&](auto const col_idx) {
+      data_page_mask_tasks.emplace_back(cudf::detail::host_worker_pool().submit_task([&] {
+        // Construct a row indices mapping based on page row counts and offsets
+        auto const total_pages_in_this_column = page_row_counts[col_idx].size();
+
+        auto const page_indices = make_page_indices_async(
+          page_row_counts[col_idx], page_row_offsets[col_idx], total_rows, streams[col_idx]);
+
+        // Device vector to hold page indices with at least one required row
+        rmm::device_uvector<size_type> select_page_indices(total_rows, streams[col_idx], mr);
+
+        // Copy page indices with at least one required row
+        auto const filtered_pages_end_iter = thrust::copy_if(
+          rmm::exec_policy_nosync(streams[col_idx]),
+          page_indices.begin(),
+          page_indices.end(),
+          thrust::counting_iterator<size_type>(0),
+          select_page_indices.begin(),
+          is_row_required_fn{row_mask.nullable(), row_mask.null_mask(), row_mask.data<bool>()});
+
+        // Remove duplicate page indices across (presorted) rows
+        auto const filtered_uniq_page_end_iter =
+          thrust::unique(rmm::exec_policy_nosync(streams[col_idx]),
+                         select_page_indices.begin(),
+                         filtered_pages_end_iter);
+
+        // Number of final filtered pages for this column
+        size_t const num_surviving_pages_this_column =
+          thrust::distance(select_page_indices.begin(), filtered_uniq_page_end_iter);
+
+        total_surviving_pages += num_surviving_pages_this_column;
+
+        // Copy the filtered page indices for this column to host
+        auto host_select_page_indices = cudf::detail::make_host_vector(
+          cudf::device_span<cudf::size_type const>{select_page_indices.data(),
+                                                   num_surviving_pages_this_column},
+          streams[col_idx]);
+
+        // Vector to data page mask the this column
+        auto valid_pages = thrust::host_vector<bool>(total_pages_in_this_column, false);
+        std::for_each(host_select_page_indices.begin(),
+                      host_select_page_indices.end(),
+                      [&](auto const page_idx) { valid_pages[page_idx] = true; });
+
+        return valid_pages;
+      }));
+    });
+
+  // Synchronize the streams
+  cudf::detail::join_streams(streams, stream);
```

Comment:
same here; tasks don't have any async cuda work and we don't do anything on the main stream until we get all task results.